File: 763.partition-labels.py
#
# @lc app=leetcode id=763 lang=python3
#
# [763] Partition Labels
#

# @lc code=start
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        letter_dict = {}
        for i in range(0, len(s)):
            if s[i] not in letter_dict.keys():
                letter_dict[s[i]] = [i, i]
            else:
                letter_dict[s[i]][1] = i


        letter_interval = list(letter_dict.values())
        letter_interval.sort(key=lambda x: x[0])


        right_interval = letter_interval[0][1]
        marker_list = []

        for i in range(1, len(letter_interval)):
            if letter_interval[i][0] <= right_interval:
                if letter_interval[i][1] > right_interval:
                    right_interval = letter_interval[i][1]
            else:
                marker_list.append(right_interval)
                right_interval = letter_interval[i][1]
            
        
        marker_list.append(len(s) - 1)
        marker_list.insert(0, -1)
        
        letter_parts = []

        for i in range(1, len(marker_list)):
            letter_parts.append(marker_list[i] - marker_list[i - 1])


        return letter_parts
        
# @lc code=end

File: 763.partition-labels.swift
/*
 * @lc app=leetcode id=763 lang=swift
 *
 * [763] Partition Labels
 */

// @lc code=start
class Solution {
    func partitionLabels(_ s: String) -> [Int] {
        struct CharInfo {
            var firstIndex: Int
            var lastIndex: Int
        }

        var charMap = [Character: CharInfo]()
        for (index, c) in s.enumerated() {
            if !charMap.keys.contains(c) {
                charMap[c] = CharInfo(firstIndex: index, lastIndex: index)
            } else {
                charMap[c]?.lastIndex = index
            }
        }

        var intervals = charMap.values.sorted {
            if $0.firstIndex == $1.firstIndex {
                return $0.lastIndex < $1.lastIndex
            } else {
                return $0.firstIndex < $1.firstIndex
            }
        }

        var end = intervals[0].lastIndex
        var result = [Int]()

        for interval in intervals[1...] {
            if interval.firstIndex < end {
                end = max(end, interval.lastIndex)
            } else {
                result.append(end)
                end = interval.lastIndex
            }
        }

        result.insert(-1, at: 0)
        result.append(s.count - 1)

        return zip(result.dropFirst(), result.dropLast()).map { $0 - $1 }

    }
}
// @lc code=end

